functorch/experimental/_map.py

```diff
@@ -17,23 +18,77 @@
     _get_current_dispatch_mode,
     _pop_mode_temporarily,
 )
-from torch.utils._pytree import tree_flatten
 from ._cond import _has_potential_branch_input_alias, _has_potential_branch_input_mutation, UnsupportedAliasMutationException
 
 
-map = HigherOrderOperator("map")
-
-
-def trace_map(proxy_mode, func_overload, f, xs, *args):
-    if not isinstance(xs, torch.Tensor):
-        raise ValueError("map() must loop over a tensor")
-    if len(xs.shape) == 0 or xs.shape[0] == 0:
-        raise ValueError("map() cannot be traced with scalar tensors or zero dimension tensors")
-    if not all(isinstance(o, torch.Tensor) for o in args):
-        raise ValueError("map() operands must be a list of tensors or modules")
+# TODO: We add this to prevent dymamo tracing into map_wrapper,
+# remove the wrapper call when it's ready.
+class MapWrapper(HigherOrderOperator):
+    def __call__(self, *args, **kwargs):
+        return map_wrapper(*args, **kwargs)
+
+map = MapWrapper("map")
+map_impl = HigherOrderOperator("map_impl")
+
+def map_wrapper(f, xs, *args):
+    flat_xs, xs_spec = pytree.tree_flatten(xs)
+    assert all(isinstance(t, torch.Tensor) for t in flat_xs), f"mapped xs can only contain tensors got {flat_xs}"
+
+    num_mapped_args = len(flat_xs)
+    assert num_mapped_args > 0, "map must have at least one mapped argument."
+
+    shapes = [xs.shape for xs in flat_xs]
+    leading_dim_size = shapes[0][0]
+    assert len(shapes) > 0 and all(cur_shape[0] == leading_dim_size for cur_shape in shapes)
+
+    out_spec = [None]
+
+    def flat_fn(num_mapped_args, *flat_args):
+        xs = pytree.tree_unflatten(flat_args[:num_mapped_args], xs_spec)
+        unflattened_out = f(xs, *flat_args[num_mapped_args:])
+        flat_out, tmp_out_spec = pytree.tree_flatten(unflattened_out)
+        out_spec[0] = tmp_out_spec
+        return flat_out
```

Comment:
Does `num_mapped_args` need to be an input to flat_fn? Removing it would make the calling convention simpler (but it's unclear to me if the design depends on it): flat_fn is the original function, except it takes a flat list of inputs. Instead of "flat_fn is the original function, except it takes in a list of flat inputs and the number of args that gets mapped over).

Reply:
I think that's a good point. Just to clarify, we won't change the signature of map_impl, right? Only omitting the num_mapped_args for flat_fn.